core/vm: implement full EOF suite [WiP]

ziogaschr opened the pull request:
This PR fetches the changes done on https://github.com/ethereum/go-ethereum/pull/26133.

gumb0 committed:
core/vm: implement EOF EIP-3540: EOF v1

chfast committed:
core/vm: simplify error handling in interpreter loop (#23952)

* core/vm: break loop on any error

* core/vm: move ErrExecutionReverted to opRevert()

* core/vm: use "stop token" to stop the loop

* core/vm: unconditionally pc++ in the loop

* core/vm: set return data in instruction impls

chfast committed:
core/vm: fill gaps in jump table with opUndefined (#24031)

gumb0 committed:
core/vm: implement EOF EIP-3670: Code Validation

gumb0 committed:
core/vm: Make INVALID a defined opcode (#24017)

* core/vm: Define 0xfe opcode as INVALID

* core/vm: Remove opInvalid as opUndefined handles it

Co-authored-by: Alex Beregszaszi <alex@rtfs.hu>
# Conflicts:
#	eth/tracers/internal/tracetest/testdata/call_tracer_legacy/inner_throw_outer_revert.json

lightclient committed:
core/vm: implement EOF EIP-4200: Static relative jumps

lightclient committed:
core/vm: implement EOF EIP-4750: Functions

# Conflicts:
#	core/vm/analysis.go
#	core/vm/contract.go
#	core/vm/evm.go
#	core/vm/instructions.go
#	core/vm/instructions_test.go
#	core/vm/interpreter.go
#	core/vm/jump_table.go
#	core/vm/stack.go

lightclient committed:
core/vm: rework eof tests

lightclient committed:
core/vm: test eof container object

lightclient committed:
core/vm: refactor eof structures

lightclient committed:
core/vm: simplify eof init in evm

lightclient committed:
core/vm: refactor instruction validator to use switch

holiman committed:
core/vm: remove go1.19 dependency

holiman committed:
core/vm: change the eof tests a bit

lightclient committed:
core/vm: refactor eof objects

lightclient committed:
core/vm: fix off-by-one in type section parser

lightclient committed:
core/vm: simplify eof parsing interface

lightclient committed:
core/vm: refactor eof more, move validation to separate file

lightclient committed:
core/vm: misc clean up

lightclient committed:
core/vm: use parse arg helper

lightclient committed:
tests: add shanghai to tests list

lightclient committed:
core/vm: don't validate code if EOF parse fails

lightclient committed:
core/vm: update callf and retf ops to b0 and b1

lightclient committed:
core/vm: ensure type section size is 2n code sections

lightclient committed:
core/vm: properly return error during contract deploy

ziogaschr committed:
cmd/hack: temp fix build by breaking interpreter loop [WiP]

ziogaschr committed:
core/vm: handle multiple interpreters on evm EOF invocations

ziogaschr committed:
core/vm: fix merge issues

ziogaschr committed:
Merge branch 'devel' into eof

ziogaschr commented:
Thinks to be done:

- [x] verify code removal at https://github.com/ETCCooperative/erigon/pull/4/commits/c5a912298f6ac6e8096087df8429171faed9cbc6
- [ ] search in code for `TODO CZ:` and verify the logic

ziogaschr commented on core/vm/absint_cfg_proof_check.go:
```diff
@@ -87,7 +87,7 @@ func resolveCheck(sem *CfgAbsSem, code []byte, st0 *astate, pc0 int) (map[int]bo
 	succs := make(map[int]bool)
 	jumps := make(map[int]bool)
 
-	if opsem == nil || opsem.halts || opsem.reverts {
```
This removal has to be verified

meowsbits replied:
LGTM. 

- `CfgOpSem` drops these fields, and there are no remaining usages of them. 
- The condition `opsem == nil` is a valid  and sufficient condition for returning because `toProgram` defines `stmt.ends = stmt.operation == nil`; so the expectation is consistent. https://github.com/ETCCooperative/erigon/pull/4/files#diff-632b254546eb0c76269d92705508d825a006ac5446f5d59dca3cfc13791c64b2R90

ziogaschr commented on core/vm/absint_cfg_proof_gen.go:
```diff
@@ -87,7 +87,7 @@ func toProgram(code []byte) *Program {
 		op := OpCode(code[pc])
 		stmt.opcode = op
 		stmt.operation = jt[op]
-		stmt.ends = stmt.operation == nil || stmt.operation.halts || stmt.operation.reverts
```
This removal has to be verified

meowsbits replied:
LGTM; ditto https://github.com/ETCCooperative/erigon/pull/4#discussion_r1037179242

ziogaschr commented on core/vm/analysis.go:
```diff
@@ -26,6 +26,14 @@ func codeBitmap(code []byte) []uint64 {
+		// Short circruit for now on EOF ops with immediates.
+		// TODO(matt): make EOF-specific code bitmap
+		if op == RJUMP || op == RJUMPI || op == CALLF {
+			// TODO CZ: verify that this is correct
```
Important to be checked as it has not been tested

ziogaschr replied:
Fixed the code with https://github.com/ETCCooperative/erigon/pull/4/commits/4ea3d3c1ba09c041a1199d409039d7877d493499. 

I think it might be able to receive some optimisations

ziogaschr replied:
Re-fixed the code so as all tests are passing as well with https://github.com/ETCCooperative/erigon/pull/4/commits/513d63f1ee334e93d6235c92828f0b4e6c479f87.

This also applies @holiman's suggestion https://github.com/ethereum/go-ethereum/pull/26133/files#r1026320917

ziogaschr commented on core/vm/errors.go:
```diff
@@ -40,8 +40,42 @@ var (
 	ErrNonceUintOverflow        = errors.New("nonce uint64 overflow")
+	ErrInvalidCode              = errors.New("invalid code: must not begin with 0xef")
+	ErrInvalidEOFCode           = errors.New("invalid code: EOF validation failed")
+	ErrInvalidInterpreter       = errors.New("invalid interpreter")
```
This new error has been added by me

meowsbits replied:
The need for this arises because erigon is still using the `Interpreter` interface, ostensibly to support external VMs. This was removed at ethereum/go-ethereum, which, with dropping configurable-EVM support, dropped the interface entirely. 

However, I'm not sure erigon's current approach is going to work with external VMs; I would expect the type assertion to fail when using, say `--vm.evm=path/to/lib.so`.
Although I'm not sure, either, that they're actually intent on keeping these configuration options, since it appears EVMC support has been dropped too. 

So we can either
- aim to drop `type Interpreter interface` and have the `evm` only handle its native self, avoiding any type assertion here
- or adapt, somehow, the implementation so that it won't fail `ErrInvalidInterpreter` when the interpreter is not of the native type `*EVMInterpreter`.

ziogaschr replied:
Nice write up. I reached the same outcome, though with the EOF PR, I was thinking that we want to deal with only the EOF surface without touching other stuff.
I suggest doing separate PRs for other changes.
What do you think?

meowsbits replied:
I think we should leave it as-is too, just making a review note of it when we do the PR.

ziogaschr commented on core/vm/evm.go:
```diff
@@ -335,6 +338,15 @@ func (evm *EVM) create(caller ContractRef, codeAndHash *codeAndHash, gas uint64,
 		return nil, common.Address{}, 0, err
 	}
+	// Check whether the init code size has been exceeded.
+	if evm.config.HasEip3860(evm.chainRules) && len(codeAndHash.code) > params.MaxInitCodeSize {
```
Wonder if brought in, more EIPs with this merge. Have to check as EIP3860 is not one of the EOF ones

meowsbits replied:
LGTM.

The checklist doc cites EIP-3860 as a prerequisite for EOF, and the EIP itself cites EIP-3670 (EOF - Code Validation) at a motivation:

> 2. To have a cost system which is extendable in the future (i.e. for proposals like [EIP-3670](https://eips.ethereum.org/EIPS/eip-3670)).

https://notes.ethereum.org/@ipsilon/eof1-checklist# 
https://eips.ethereum.org/EIPS/eip-3860

ziogaschr commented on core/vm/instructions.go:
```diff
@@ -889,36 +905,40 @@ func makeLog(size int) executionFunc {
+			pcAbsolute = context.CodeOffset + *pc
+			startMin   = pcAbsolute + 1
+		)
+		if startMin >= codeEnd {
```
the conditionals logic here is the exact opposite from go-ethereum one. We have to check if this was on purpose or just never merged from upstream. Maintained the erigon's principle for now.

meowsbits replied:
I just move this scope to use upstream logic with ed99df442fffa33e9224e9bc2454578466986d4e because I don't see any reason in the history why it diverged from upstream; the latest meaningful change here was from the EOF work merge commit.

ziogaschr replied:
Can you have a look at https://github.com/ledgerwatch/erigon/commit/199ede36a2661908343004264ba304fce31a8793#diff-7da47ad076053fe0f266b3088eb9ee206c7f3f8fe5a93116a3f5e25ee5074213.
Both are at Erigon's repo

meowsbits replied:
Good find, thanks! 
By my estimation, we can leave the logic as-is (now) without negatively impacting the changes that that PR/patch introduced because 
- my patch (https://github.com/ETCCooperative/erigon/pull/4/commits/ed99df442fffa33e9224e9bc2454578466986d4e)  doesn't change anything with slice allocations
- and the `RightPadBytes` vs. `SetBytes` approach/patch wasn't ultimately adopted; or at least does not seem to be in the original codebase this is on top of
  + https://github.com/ETCCooperative/erigon/pull/4/files#diff-7da47ad076053fe0f266b3088eb9ee206c7f3f8fe5a93116a3f5e25ee5074213L920 (see below for ?)

Something seems wonky; 
- https://github.com/ledgerwatch/erigon/pull/638/files accepts `SetBytes` over `RightPadBytes`, but `RightPadBytes` is still the form at our base ref...?

meowsbits replied:
The PR/patch introducing `uint256` seems to be the culprit which reverted the removal of `RightPadBytes` - https://github.com/ledgerwatch/erigon/commit/924c03abf999989373684902540b8b94cbb0a1ad#diff-7da47ad076053fe0f266b3088eb9ee206c7f3f8fe5a93116a3f5e25ee5074213L879-R929.

ziogaschr replied:
Yes, this is why I noted this part of code to be reviewed, as it was a bit annoying to check the final intentions and outcome.
For this reason, I decided to leave this as is in Erigon's side.
What do you think?

meowsbits replied:
I've just reverted my change that moved the logic to go-ethereum-style; now the function looks as erigon had it originally, but, of course, handling the EOF CodeOffset, CodeLength parameters. 
https://github.com/ETCCooperative/erigon/pull/4/commits/2f4a71742724df4853d67a60a5b767fc17d39733

ziogaschr commented on core/vm/interpreter.go:
```diff
@@ -66,9 +66,20 @@ type Interpreter interface {
-	Stack    *stack.Stack
-	Contract *Contract
+	Memory        *Memory
+	Stack         *stack.Stack
```
Go-ethereum is using [Stack](https://github.com/ethereum/go-ethereum/blob/b3f36e10766956bf204f2d2b9415dcd8cfd2be6b/core/vm/stack.go). Probably at some point they moved the code and Erigon hasn't followed.

meowsbits replied:
Are you thinking that moving to a homegrown `Stack` is outside the scope of this changeset?

ziogaschr replied:
Not sure,  but thinking that it might be easier for Erigon team to review only the EOF parts. Later we can make PRs that change stuff like `Stack`.
What do you think?

meowsbits replied:
Yep, SGTM. Let's consider it out of scope.

ziogaschr commented on core/vm/jump_table.go:
```diff
@@ -86,6 +90,8 @@ func newShanghaiInstructionSet() JumpTable {
 	enable3855(&instructionSet) // PUSH0 instruction https://eips.ethereum.org/EIPS/eip-3855
 	enable3860(&instructionSet) // Limit and meter initcode https://eips.ethereum.org/EIPS/eip-3860
+	enable4200(&instructionSet) // Static relative jumps https://eips.ethereum.org/EIPS/eip-4200
+	enable4750(&instructionSet) // Functions https://eips.ethereum.org/EIPS/eip-4750
```
Check and activate the appropriate EIPs for Shanghai

meowsbits replied:
- [x] 3855 is included in Shanghai, but not for EOF
- [x] 3860 is included in Shanghai and necessary for EOF, but not exclusive to EOF

---

EOF-specific EIPs (are any missing?)

- [x] 3540 (EVM Object Format v1) :heavy_check_mark: Turns on format validation, but has no modifications for the instruction set.
- [x] 3670 (EOF - Code Validation) :heavy_check_mark: Implemented, no need for mod at the instruction set.
- [x] 4200 (EOF - Static Jumps) :heavy_check_mark: 
- [x] 4750 (EOF - Functions) :heavy_check_mark:  
- [ ] 5450 (Stack Validation) :question: Implementation? Depends on 4750 (EOF - Functions).

ziogaschr replied:
EIP-5450 is on the works here https://github.com/lightclient/go-ethereum/tree/eof-5450

ziogaschr commented on core/vm/opcodes.go:
```diff
@@ -200,11 +202,10 @@ const (
 	LOG4
 )
 
-// unofficial opcodes used for parsing.
```
I removed those unofficial opcodes and they seem as leftovers. Let's verify they truly are

meowsbits replied:
LGTM.

These "unofficial opcodes used for parsing" were in the `0xb0` range. They have been replaced with "control flow ops", namely `CALLF` and `RETF`, both ops via EIP-4750 (EOF - Functions). 

The original ops `PUSH`, `DUP`, and `SWAP` were shorthands for stack ops using scalar values (like `PUSH1`, `PUSH2`, etc.), and they are not used anywhere anymore.

ziogaschr commented on core/vm/validation.go:
```diff
@@ -0,0 +1,79 @@
+			// Check if offset points to non-code segment.
+			// TODO(matt): include CALLF and RJUMPs in analysis.
+			if analysis == nil {
+				// TODO CZ: implement the logic https://github.com/ethereum/go-ethereum/pull/26133/commits/370c24c344f7fc26cc29fba9b33825baeded8876#diff-882af11429522494f308c57017fe1c44e86b338e3bf5c308584d9d187606d252L64
```
This has to be implemented

ziogaschr replied:
resolved https://github.com/ETCCooperative/erigon/pull/4/commits/3b35f1d5602b9aaac5b702c486d9e7b64a1a92b6

ziogaschr committed:
core/vm: remove OpTstore test as it's not implemented yet

ziogaschr committed:
core/vm: fix lint issues after merge

ziogaschr committed:
core/vm: handle analysis init on validateCode

this logic had to be checked after merge of EOF from upstream
https://github.com/ETCCooperative/erigon/pull/4#discussion_r1036252354

ziogaschr committed:
core/vm: fix codeBitmap logic

ziogaschr committed:
core/vm: refix codeBitmap based on holiman’s suggestion

ziogaschr commented:
At this point tests are passing.
Still have to review code, comments in this PR and keep fetching updates from https://github.com/ethereum/go-ethereum/pull/26133 as it is WiP.

meowsbits committed:
core/vm: rename eof1 -> eof1Only for consistency with legacyOnly

Date: 2022-12-01 06:11:09-08:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
core/vm: use ethereum/go-ethereum conditional logic for push constructor

This (re)uses the exact same logic as ethereum/go-ethereum
at b3f36e10766956bf204f2d2b9415dcd8cfd2be6b.

Rel https://github.com/ETCCooperative/erigon/pull/4#pullrequestreview-1199734442

The latest 'blame' on this stanza shows that the latest change
resulted from the merge commit, so I assume that
logic in here is not intentionally custom or reactionary
to the upstream implementation.

Date: 2022-12-01 06:50:22-08:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
core/vm: add sanity test for opcode::string maps

Date: 2022-12-01 07:36:37-08:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr commented on core/vm/opcodes_test.go:
```diff
@@ -0,0 +1,15 @@
+)
+
+// TestOpcodeStringer is a sanity check to test the internal consistency of our opcode:string mappings.
+func TestOpcodeStringer(t *testing.T) {
```
That's a nice little test :) Thanks

meowsbits committed:
core/vm: remove unused and commented DUP,SWAP string assignments

Date: 2022-12-01 07:49:08-08:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits reviewed (commented):
https://github.com/ETCCooperative/erigon/pull/4/commits/e0de7760619fa88a673b1fdb37230f2de9223dd2

This is not 1:1 with ethereum/go-ethereum; my heart will not be broken if we choose to revert this for the sake of minimizing the diff.

meowsbits committed:
core/vm: remove commented operation fields .reverts and .halts

Date: 2022-12-01 10:03:34-08:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr commented:
> [e0de776](https://github.com/ETCCooperative/erigon/commit/e0de7760619fa88a673b1fdb37230f2de9223dd2)
> 
> This is not 1:1 with ethereum/go-ethereum; my heart will not be broken if we choose to revert this for the sake of minimizing the diff.



Your change makes sense to me. Up to you to decide if we want to stay close on upstream PR or suggest more. This is so tiny change, that it's fine to keep I think.

meowsbits committed:
Revert "core/vm: use ethereum/go-ethereum conditional logic for push constructor"

This reverts commit ed99df442fffa33e9224e9bc2454578466986d4e.

meowsbits reviewed (commented):
CI `Lint` step is failing: https://github.com/ETCCooperative/erigon/actions/runs/3609205418/jobs/6082252376
But can't reproduce the failure locally :confused: 

```
uos 22-12-03T08:04:31 ~/dev/ledgerwatch/erigon eof=
make lintci-deps && make lintci && make lint
rm -f ./build/bin/golangci-lint
curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b ./build/bin v1.50.1
golangci/golangci-lint info checking GitHub for tag 'v1.50.1'
golangci/golangci-lint info found version: 1.50.1 for v1.50.1/linux/amd64
golangci/golangci-lint info installed ./build/bin/golangci-lint
--> Running linter for code
```

meowsbits commented on core/vm/analysis.go:
```diff
@@ -26,6 +26,20 @@ func codeBitmap(code []byte) []uint64 {
+		// Short circruit for now on EOF ops with immediates.
+		// TODO(matt): make EOF-specific code bitmap
+		if op == RJUMP || op == RJUMPI || op == CALLF {
+			// TODO(CZ): For now the logic of bitvec.setN has been copied
```
@ziogaschr Is this still a TODO? 

Ref https://github.com/ethereum/go-ethereum/pull/26133/files#diff-882af11429522494f308c57017fe1c44e86b338e3bf5c308584d9d187606d252R75

ziogaschr commented:
It is working. Though what do you think, shall we abstract it in this PR or leave it there as it has been removed in Erigon?
On 5 Dec 2022, 16:50 +0200, meowsbits ***@***.***>, wrote:
> @meowsbits commented on this pull request.
> In core/vm/analysis.go:
> > @@ -26,6 +26,20 @@ func codeBitmap(code []byte) []uint64 {
>        for pc := 0; pc < len(code); {
>                op := OpCode(code[pc])
>                pc++
> +               // Short circruit for now on EOF ops with immediates.
> +               // TODO(matt): make EOF-specific code bitmap
> +               if op == RJUMP || op == RJUMPI || op == CALLF {
> +                       // TODO(CZ): For now the logic of bitvec.setN has been copied
> @ziogaschr Is this still a TODO?
> Ref https://github.com/ethereum/go-ethereum/pull/26133/files#diff-882af11429522494f308c57017fe1c44e86b338e3bf5c308584d9d187606d252R75
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>

meowsbits committed:
Merge branch 'devel' into eof

Conflicts:
  core/vm/interpreter.go

meowsbits committed:
core/vm: use correct pc var

Date: 2022-12-05 09:48:23-08:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits commented:
New thread: I just merged ledgerwatch/erigon@devel with [6e2b198](https://github.com/ETCCooperative/erigon/pull/4/commits/6e2b19837f7e472fb46cb5bbd7c8978637878ee1), PTAL; there were a few conflicts I resolved in core/vm/interpreter.go.

ziogaschr committed:
core/vm: switch dep of “github.com/ethereum/go-ethereum/common” to “github.com/ledgerwatch/erigon/common”

ziogaschr committed:
core/vm: remove superfluous code

ziogaschr committed:
Merge branch 'devel' of github.com:ledgerwatch/erigon into eof

# Conflicts:
#	go.mod

ziogaschr committed:
tests: fix tests by /s/ShanghaiBlock/ShanghaiTime

ziogaschr committed:
tests: keep ArrowGlacierToMergeAtDiffC0000 as a fork in tests

ziogaschr committed:
tests: revert head for testdata to 9e058e5

ziogaschr committed:
Update core/vm/eips.go

Co-authored-by: Andrew Ashikhmin <34320705+yperbasis@users.noreply.github.com>

ziogaschr committed:
core/vm: mark `CALLCODE` & `SELFDESTRUCT` as legacyOnly based on EIP3670

ziogaschr committed:
common/compiler: fix vyper test

ziogaschr committed:
core/vm: remove superfluous check

ziogaschr commented:
Closing in favour of PR https://github.com/ledgerwatch/erigon/pull/6382.
